### PR TITLE
Add initial support for file batch DDL sqlcmd feature

### PR DESCRIPTION
**Pull request by jxchinaren**
For CORE team review

### CODE REVIEW

File: src/frontend/org/voltdb/parser/SQLParser.java [hunk 1]
```diff
@@ -876,17 +877,33 @@ public static ParseRecallResults parseRecallStatement(String statement, int line
...
-    public static File parseFileStatement(String statement)
+    public static FileInfo parseFileStatement(String statement)
     {
-        Matcher matcher = FileToken.matcher(statement);
-        if (matcher.matches()) {
-            return new File(matcher.group(1));
+        Matcher batchMatcher = FileBatchToken.matcher(statement);
+        if (batchMatcher.matches()) {
+            return new FileInfo(new File(batchMatcher.group(1)), true);
+        } else {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/parser/SQLParser.java]: eliminate else after a return.

File: src/frontend/org/voltdb/utils/SQLCommand.java [hunk 2]
```diff
@@ -236,6 +271,29 @@ private static String handleTranslatedCommands(String lineIn)
         return lineIn;
     }
 
+    public static class QueryInfo {
+        StringBuilder query;
+        boolean batch = false;
```
Comment (wolffcm) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: I could envision this class as having more members and used in more places in the future.  It would be good to make the members private and provide accessors.  Also, can this class be made immutable?  (These are both items in the Effective Java book.)

File: src/frontend/org/voltdb/utils/SQLCommand.java [hunk 3]
```diff
@@ -236,6 +271,29 @@ private static String handleTranslatedCommands(String lineIn)
         return lineIn;
     }
 
+    public static class QueryInfo {
+        StringBuilder query;
```
Comment (wolffcm) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: We seem to use the word "query" pretty freely in our code to refer to any SQL statement, but really it refers to statements (like SELECT) that return rows from the database.  Since I think you mean to include DDL statements too here, I think `StmtInfo` would be a better name for this class?

Comment (jxchinaren) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: Did you see that we use "query" variable as a StringBuilder all over to concat all the input DDL/Queries/Directives, etc. I guess I put a blind eye on it and follow it.

I would love to do a refactor work as a separate commit to fix it.

Comment (wolffcm) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: Sounds good, thanks.

File: src/frontend/org/voltdb/parser/SQLParser.java [hunk 4]
```diff
@@ -876,17 +877,40 @@ public static ParseRecallResults parseRecallStatement(String statement, int line
...
+    public static FileInfo parseFileStatement(String statement)
     {
+        Matcher batchMatcher = FileBatchToken.matcher(statement);
+        if (batchMatcher.matches()) {
+            return new FileInfo(new File(batchMatcher.group(1)), true);
+        }
         Matcher matcher = FileToken.matcher(statement);
         if (matcher.matches()) {
-            return new File(matcher.group(1));
+            return new FileInfo(new File(matcher.group(1)), false);
```
Comment (paulmartel) [inline src/frontend/org/voltdb/parser/SQLParser.java]: minor: you could factor the "new File(filepath)" step into the FileInfo constructor.

File: src/frontend/org/voltdb/utils/SQLCommand.java [hunk 5]
```diff
@@ -287,7 +360,7 @@ public static void executeNoninteractive() throws Exception
     /// input loop can proceed to the next line.
     //TODO: There have been suggestions that some or all of these directives could be made
     // available in non-interactive contexts. This function is available to enable that.
-    private static boolean executesAsSimpleDirective(String line) throws Exception {
+    private static boolean executesShowCmdAsSimpleDirective(String line) throws Exception {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: This renaming is wrong -- the function also implements the Help directive -- revert the name.

File: src/frontend/org/voltdb/utils/SQLCommand.java [hunk 6]
```diff
@@ -434,7 +507,24 @@ private static void printTables(final String name, final Collection<String> tabl
...
     }
 
-    public static String readScriptFile(File file)
+    // function update the queryBatchList and query
+    private static StringBuilder updateQueryBatch(List<QueryInfo> contentInfo,
+            List<QueryInfo> queryBatchList, StringBuilder query) {
+        if (m_inFileBatchMode) {
+            assert(contentInfo.size() == 1);
+            // found contents for FILE cmd, flush the previous lines into a batch
+            queryBatchList.add(new QueryInfo(query, false));
```
Comment (paulmartel) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: Should we skip this QueryInfo entry if query.isEmpty()?

File: src/frontend/org/voltdb/utils/SQLCommand.java [hunk 7]
```diff
@@ -434,7 +507,24 @@ private static void printTables(final String name, final Collection<String> tabl
...
+            query = new StringBuilder();
+
+            queryBatchList.add(new QueryInfo(contentInfo.get(0).getQuery(), true));
+        } else {
+            queryBatchList.addAll(contentInfo);
+        }
+
+        return query;
```
Comment (paulmartel) [inline src/frontend/org/voltdb/utils/SQLCommand.java]: It does not seem right in non-batch mode to "place on hold" the current contents of query while we queue batch contents before it.
Would it be better to always queue query in this function, have no return value, and let the callers always reset the StringBuilder?

File: tests/frontend/org/voltdb/utils/TestSqlCmdErrorHandling.java [hunk 8]
```diff
@@ -189,12 +189,18 @@ private int callSQLcmdBase(boolean stopOnError, String inputText,
...
-            pb = new ProcessBuilder(commandPath, "--ddl-file=" + f.getPath());
+        if (batchModeDDL) {
+            pb = new ProcessBuilder(commandPath);
+            File batchScript = new File("script.sql");
+            batchScript.deleteOnExit();
+            fos = new FileOutputStream(batchScript);
+            fos.write(new String("FILE -batch " + f.getPath()).getBytes());
+            fos.close();
+
+            pb.redirectInput(batchScript);
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/utils/TestSqlCmdErrorHandling.java]: I like having a few basic batch mode ddl unit tests here.

I don't remember any plans to deprecate all testing for "--ddl-file=".
Maybe these test cases should exist in parallel?

**Comment by paulmartel** [comment]
End of review round.
